src/geometry/cuda/edge_coarsening_impl.cu

```diff
@@ -0,0 +1,261 @@
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file geometry/cuda/edge_coarsening_impl.cu
+ * \brief Edge coarsening CUDA implementation
+ */
+#include <dgl/array.h>
+#include <curand.h>
+#include <dgl/random.h>
+#include <cstdint>
+#include "../geometry_op.h"
+#include "../../runtime/cuda/cuda_common.h"
+
+#define THREADS 1024
+#define BLOCKS(N) (N + THREADS - 1) / THREADS
+#define BLUE_P 0.53406
+#define BLUE -1
+#define RED -2
+#define EMPTY_IDX -1
+#define CURAND_CALL(func)           \
+{                                   \
+  curandStatus_t e = (func);        \
+  CHECK(e == CURAND_STATUS_SUCCESS) \
+    << "CURAND: Error at "          \
+    << __FILE__ << ":" << __LINE__; \
+}
+
+namespace dgl {
+namespace geometry {
+namespace impl {
+
+__device__ bool done_d;
+__global__ void init_done_kernel() { done_d = true; }
+
+template <typename IdType>
+__global__ void colorize_kernel(const float *prop, int64_t num_elem, IdType *result) {
+  const IdType idx = blockIdx.x * blockDim.x + threadIdx.x;
+  if (idx < num_elem) {
+    if (result[idx] < 0) {  // if unmatched
+      result[idx] = (prop[idx] > BLUE_P) ? RED : BLUE;
+      done_d = false;
+    }
+  }
+}
+
+template <typename IdType>
+__global__ void propose_kernel(const IdType *indptr, const IdType *indices,
+                               int64_t num_elem, IdType *proposal, IdType *result) {
+  const IdType idx = blockIdx.x * blockDim.x + threadIdx.x;
+  if (idx < num_elem) {
+    if (result[idx] != BLUE) return;
+
+    bool has_unmatched_neighbor = false;
+    for (IdType i = indptr[idx]; i < indptr[idx + 1]; ++i) {
+      auto v = indices[i];
+
+      if (result[v] < 0)
+        has_unmatched_neighbor = true;
+      if (result[v] == RED) {  // propose to the first red neighbor
+        proposal[idx] = v;
+        break;
+      }
+    }
+    __syncthreads();
```

Comment:
This acts as a barrier but only works for threads inside a block, and there might be multiple blocks executed in parallel. We can not fully avoid RAW (read after write) in this way.
Try separating read and write: `propose_kernel(..., const IdType *result_in, IdType *result_out)`.
btw, do we need to avoid race condition here?

Reply:
The description in the original paper does not mention avoiding race condition and the implementation in PyG does not have the `__syncthreads()` here. I will remove this.